xarray/core/dataset.py

```diff
@@ -210,6 +224,7 @@ def __init__(self, data_vars=None, coords=None, attrs=None,
             coords = {}
         if data_vars is not None or coords is not None:
             self._set_init_vars_and_dims(data_vars, coords, compat)
+            self._check_multiindex_level_names()
```

Comment:
OK I see, `Dataset._set_init_vars_and_dims` indeed depends on `merge.py` but I actually didn't go more into that module. I'll take a look.